pennylane/gradients/pulse_generator_gradient.py

```diff
@@ -234,3 +250,209 @@ def _generate_tapes_and_coeffs(tape, idx, atol, cache):
+        single_measure (bool): whether the tape execution results contain single measurements.
+        shot_vector (bool): whether the tape execution results were obtained with a shot vector.
+
+    Returns:
+        tensor_like or tuple[tensor_like] or tuple[tuple[tensor_like]]: contraction between the
+        parameter-shift derivative computed from ``results`` and the ``coeffs``.
+    """
+
+    def _parshift_and_contract_single(res_list, coeffs):
+        """Execute the standard parameter-shift rule on a list of results
+        and contract with Pauli basis coefficients."""
+        psr_deriv = ((res := qml.math.stack(res_list))[::2] - res[1::2]) / 2
+        return qml.math.tensordot(psr_deriv, coeffs, axes=[[0], [0]])
+
+    if single_measure and not shot_vector:
+        # single measurement and single shot entry
+        return _parshift_and_contract_single(results, qml.math.stack(coeffs))
+    if single_measure or not shot_vector:
+        # single measurement or single shot entry, but not both
+        return tuple(
+            _parshift_and_contract_single(r, qml.math.stack(coeffs)) for r in zip(*results)
+        )
+
+    return tuple(
+        tuple(_parshift_and_contract_single(_r, qml.math.stack(coeffs)) for _r in zip(*r))
+        for r in zip(*results)
+    )
+
+
+def _expval_pulse_generator(tape, argnum, shots, atol):
+    """Compute the pulse generator parameter-shift rule for a quantum circuit that returns expectation
+    values of observables.
+
+    Args:
+        tape (`~.QuantumTape`): Quantum circuit to be differentiated with the pulse generator
+            parameter-shift rule.
+        argnum (int or list[int] or None): Trainable tape parameter indices to differentiate
+            with respect to. If not provided, the derivatives with respect to all
+            trainable parameters are returned.
+        shots (None, int, list[int]): The shots of the device used to execute the tapes which are
+            returned by this transform. Note that this argument does not *influence* the shots
+            used for execution, but *informs* the transform about the shots to ensure a compatible
+            return value formatting.
+        atol (float): absolute tolerance used to determine vanishing contributions.
+
+    Returns:
+        function or tuple[list[QuantumTape], function]:
+
+        - If the input is a QNode, an object representing the Jacobian (function) of the QNode
+          that can be executed to obtain the Jacobian.
+          The type of the Jacobian returned is either a tensor, a tuple or a
+          nested tuple depending on the nesting structure of the original QNode output.
+
+        - If the input is a tape, a tuple containing a
+          list of generated tapes, together with a post-processing
+          function to be applied to the results of the evaluated tapes
+          in order to obtain the Jacobian.
+
+    """
+    argnum = argnum or tape.trainable_params
```

Comment:
Otherwise `0` gets cast to `False` and higher integers create an error in line 347.
Also, note that `argnums=[]` results in all parameters being differentiated. This is probably not what the user expects.
```suggestion
    if isinstance(argnum, int):
        argnum = [argnum]
    argnum = argnum or tape.trainable_params
```

Reply:
Mhm. Good point. I think this behaviour exists throughout the module. I like the fix for integers,but would go with the possibly unexpected behaviour for `argnum=[]`  for now. We can change this behaviour (with a deprecation cycle) in another PR :)

Reply:
Opened #4213 .

Reply:
Thanks, I wanted to incorporate this before but somehow missed to do it in the end. Included the above now.